metaflow/plugins/kfp/kfp_split_contexts.py

```diff
@@ -0,0 +1,164 @@
+import json
+import os
+import pprint
+from typing import Callable, Dict
+
+from metaflow import S3, FlowSpec, current
+from metaflow.datastore import MetaflowDataStore
+from metaflow.graph import DAGNode, FlowGraph
+from metaflow.plugins.kfp.kfp_constants import (
+    KFP_METAFLOW_CONTEXT_DICT_PATH,
+    PASSED_IN_SPLIT_INDEXES_ENV_NAME,
+    SPLIT_SEPARATOR,
+)
+
+
+class KfpSplitContext(object):
+    """
+    passed_in_split_indexes is a string of foreach split_index ordinals.
+    A nested foreach appends the new split index ordinal with a "_" separator.
+    Example:
+        0_1 -> 0th index of outer foreach and 1th index of inner foreach
+        1_0 -> 1th index of outer foreach and 0th inde of inner foreach
```

Comment:
This index schema nicely orders output. Nice.

A small problem whenthere are more than one sets of complete split-join steps: you will end up with multiple copies of indices `0` and `1`, each from a set of splits. For quick concrete example, imagine 2 copied of the [flow example](https://github.com/zillow/metaflow/blob/tz/foreach/metaflow/plugins/kfp/metaflow_nested_foreach.ipynb) you provide is concatenated in series.

Reply:
Nice thinking! Metaflow doesn't support this scenario (thankfully, they keep it simple and the philosophy is to not allow complex flows and confuse Applied Scientists, plus it's very rare)

I tested it.

```python
from metaflow import FlowSpec, step  # type: ignore


class BranchFlow(FlowSpec):
    """
    BranchFlow documentation
    """

    @step
    def start(self):
        self.next(self.a, self.b)

    @step
    def a(self):
        self.x = 1
        print(f"self.x = {self.x}")
        self.titles = ['Stranger Things',
                       'House of Cards',
                       'Narcos']
        self.next(self.mapper, foreach='titles')

    @step
    def b(self):
        self.x = 2
        print(f"self.x = {self.x}")
        self.titles = ['Stranger Things',
                       'House of Cards',
                       'Narcos']
        self.next(self.mapper, foreach='titles')

    @step
    def mapper(self, input):
        print(f"mapper {input}")
        self.next(self.mapper_join)

    @step
    def mapper_join(self, inputs):
        self.next(self.join)

    @step
    def join(self, inputs):
        total = sum(input.x for input in inputs)
        print(f"inputs.a.x = {inputs.a.x}")
        print(f"inputs.b.x = {inputs.b.x}")
        print(f"total is {total}")

        # validate ourselves
        self.total = 3
        assert self.total == 3

        self.next(self.end)

    @step
    def end(self):
        assert self.total == 3


if __name__ == "__main__":
    BranchFlow()

```

```bash
$ python branch_flow.py
Metaflow 2.2.3.post104+git5ded15b executing BranchFlow for user:talebz@zillowgroup.com
Validating your flow...
    Validity checker found an issue on line 31:
    Step mapper seems like a join step (it takes an extra input argument) but an incorrect number of steps (a, b) lead to it. This join was expecting 1 incoming paths, starting from splitted step(s) mapper.
```